### PR TITLE
Implement `take` kernel for byte view array.

### LINKED ISSUE #5511: `take` kernel support for `StringViewArray` and `BinaryViewArray`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
This is part of the larger project to implement `StringViewArray` -- see https://github.com/apache/arrow-rs/issues/5374

In https://github.com/apache/arrow-rs/pull/5481 we added support for `StringViewArray` and `ByteViewArray`.

This ticket tracks supporting `StringViewArray` and `ByteViewArray` in the `take` kernel: https://docs.rs/arrow/latest/arrow/compute/kernels/take/index.html

**Describe the solution you'd like**
1. Support `StringViewArray` and `ByteViewArray`  the `take` kernel
2. Add tests

Notes: 
1. you should be able to follow the model of `DictionaryArray` as the idea is similar
2. Filtering should *not* copy any strings, it should just manipulate the fixed size views

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
# Which issue does this PR close?

Closes #5511.

# Rationale for this change
 

Necessary feature.

# What changes are included in this PR?

- [x] Implement `take` kernel for `StringViewArray` and `ByteViewArray`.
- [x] Add unit tests for the kernel.

# Are there any user-facing changes?

No.

### CODE REVIEW

File: arrow-select/src/take.rs [hunk 1]
```diff
@@ -437,6 +443,20 @@ fn take_bytes<T: ByteArrayType, IndexType: ArrowPrimitiveType>(
...
+/// `take` implementation for byte view arrays
+fn take_byte_view<T: ByteViewType, IndexType: ArrowPrimitiveType>(
+    array: &GenericByteViewArray<T>,
+    indices: &PrimitiveArray<IndexType>,
+) -> Result<GenericByteViewArray<T>, ArrowError> {
+    let new_views = take_native(array.views(), indices);
+    let new_nulls = take_nulls(array.nulls(), indices);
+    Ok(GenericByteViewArray::new(
+        new_views,
+        array.data_buffers().to_vec(),
```
Comment (RinChanNOWWW) [inline arrow-select/src/take.rs:455]: Should we prune the data buffers right now?

Comment (tustvold) [reply arrow-select/src/take.rs:455]: I think this should be a separate kernel to optimise

File: arrow-array/src/cast.rs [hunk 2]
```diff
@@ -779,6 +779,34 @@ pub trait AsArray: private::Sealed {
         self.as_bytes_opt().expect("binary array")
     }
 
+    /// Downcast this to a [`StringViewArray`] returning `None` if not possible
+    fn as_string_view(&self) -> &StringViewArray {
```
Comment (tustvold) [inline arrow-array/src/cast.rs:783]: This is not a breaking change as the trait is sealed

File: arrow-select/src/take.rs [hunk 3]
```diff
@@ -437,6 +443,20 @@ fn take_bytes<T: ByteArrayType, IndexType: ArrowPrimitiveType>(
     Ok(GenericByteArray::from(array_data))
 }
 
+/// `take` implementation for byte view arrays
+fn take_byte_view<T: ByteViewType, IndexType: ArrowPrimitiveType>(
+    array: &GenericByteViewArray<T>,
+    indices: &PrimitiveArray<IndexType>,
+) -> Result<GenericByteViewArray<T>, ArrowError> {
+    let new_views = take_native(array.views(), indices);
+    let new_nulls = take_nulls(array.nulls(), indices);
```
Comment (tustvold) [inline arrow-select/src/take.rs:452]: It's beautiful how these traits just magically work

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
I took the liberty of switching to use ArrayData equality and renaming the AsArray method to byte_view. Whilst `bytes` is perhaps more consistent with the `as_bytes()` method, I think it is better to be consistent with the naming of the type (`as_bytes` is probably named incorrectly)

**Comment by alamb** [comment]
Thank you @RinChanNOWWW  and @tustvold 

FYI @ariesdevil and @XiangpengHao  -- its happening!

**Pull request merged**